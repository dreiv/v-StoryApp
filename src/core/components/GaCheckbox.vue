<script setup lang="ts">
import { useTemplateRef } from 'vue'
import { CheckIcon, MinusIcon } from 'lucide-vue-next'
import { useAria, type AriaProps } from '../composables/useAria'

defineOptions({ inheritAttrs: false })
const props = defineProps<AriaProps>()

const { ariaStyles, ariaAttrs } = useAria(props)
const inputRef = useTemplateRef<HTMLElement>('inputRef')
const model = defineModel<boolean>()

defineExpose({ inputRef })
</script>

<template>
  <input
    ref="inputRef"
    type="checkbox"
    :class="[$style.native, ...ariaStyles]"
    v-model="model"
    v-bind="{ ...$attrs, ...ariaAttrs }"
  />

  <div :class="$style.marker">
    <CheckIcon :class="$style.checked" :size="12" :stroke-width="4" />
    <MinusIcon :class="$style.indeterminate" :size="12" :stroke-width="4" />
  </div>
</template>

<style module>
.native,
.marker {
  margin: var(--ga-size-spacing-01) 0 var(--ga-size-spacing-01) var(--ga-size-spacing-01);
  width: var(--ga-size-spacing-05);
  height: var(--ga-size-spacing-05);
}

.marker {
  position: absolute;
  top: 0;
  left: 0;

  border: var(--ga-size-border-width-md) solid var(--ga-color-border-action);
  border-radius: var(--ga-radius);

  pointer-events: none;
  color: var(--ga-color-icon-on-primary);

  & > .checked,
  & > .indeterminate {
    display: none;
  }
}

.native {
  flex-shrink: 0;
  opacity: 0;
  cursor: pointer;

  &:focus-visible + .marker {
    outline: var(--ga-size-border-width-md) solid var(--ga-color-border-focus);
    outline-offset: var(--ga-size-spacing-01);
  }

  &:hover:enabled {
    &:not(:checked, :indeterminate) + .marker {
      border-color: var(--ga-color-border-action-hover);
      background-color: var(--ga-color-surface-action-hover-2);
    }

    &:checked + .marker,
    &:indeterminate + .marker {
      background-color: var(--ga-color-surface-action-hover);
    }
  }

  &:checked + .marker,
  &:indeterminate + .marker {
    background-color: var(--ga-color-surface-action);
  }
  &:checked + .marker > .checked {
    display: block;
  }
  &:indeterminate + .marker > .indeterminate {
    display: block;
  }

  &:disabled {
    cursor: not-allowed;

    + .marker {
      border-color: var(--ga-color-border-disabled);
      background-color: var(--ga-color-surface-disabled);
      color: var(--ga-color-icon-on-disabled);
    }

    ~ .label {
      color: var(--ga-color-text-disabled);
    }
  }

  &.error {
    + .marker {
      border-color: var(--ga-color-border-error);
      background-color: var(--ga-color-surface-error);
      color: var(--ga-color-icon-error);
    }

    &:checked + .marker,
    &:indeterminate + .marker {
      background-color: var(--ga-color-surface-error);
    }

    &:hover:enabled + .marker {
      border-color: var(--ga-color-border-error-hover);
      background-color: var(--ga-color-surface-action-hover-2);
    }
  }
}
</style>
