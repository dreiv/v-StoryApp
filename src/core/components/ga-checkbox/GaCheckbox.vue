<script setup lang="ts">
import { useTemplateRef } from 'vue'
import { CheckIcon, MinusIcon } from 'lucide-vue-next'
import { useFormInput, type FormInputProps } from '@/core/composables/useFormInput'

defineOptions({ inheritAttrs: false })
const props = defineProps<FormInputProps>()

const { classes, aria } = useFormInput(props)
const inputRef = useTemplateRef<HTMLElement>('inputRef')
const model = defineModel<boolean>()

defineExpose({ inputRef })
</script>

<template>
  <label :class="classes" v-bind="aria">
    <input ref="inputRef" type="checkbox" :class="$style.native" v-model="model" v-bind="$attrs" />
    <div :class="$style.marker">
      <CheckIcon :class="$style.checked" :size="12" :stroke-width="4" />
      <MinusIcon :class="$style.indeterminate" :size="12" :stroke-width="4" />
    </div>

    <span :class="$style.label">
      <slot>{{ label }}</slot>
    </span>
  </label>
</template>

<style module>
.input {
  display: inline-flex;
  position: relative;
  gap: var(--ga-size-spacing-03);

  user-select: none;
}

.native,
.marker {
  margin: var(--ga-size-spacing-01) 0 var(--ga-size-spacing-01) var(--ga-size-spacing-01);
  width: var(--ga-size-spacing-05);
  height: var(--ga-size-spacing-05);
}

.marker {
  position: absolute;
  top: 0;
  left: 0;

  border: var(--ga-size-border-width-md) solid var(--ga-color-border-action);
  border-radius: var(--ga-radius);

  pointer-events: none;
  color: var(--ga-color-icon-on-primary);

  & > .checked,
  & > .indeterminate {
    display: none;
  }
}

.native {
  flex-shrink: 0;
  opacity: 0;
  cursor: pointer;

  &:focus-visible + .marker {
    outline: var(--ga-size-border-width-md) solid var(--ga-color-border-focus);
    outline-offset: var(--ga-size-spacing-01);
  }

  &:hover:enabled {
    &:not(:checked, :indeterminate) + .marker {
      border-color: var(--ga-color-border-action-hover);
      background-color: var(--ga-color-surface-action-hover-2);
    }

    &:checked + .marker,
    &:indeterminate + .marker {
      background-color: var(--ga-color-surface-action-hover);
    }
  }

  &:checked + .marker,
  &:indeterminate + .marker {
    background-color: var(--ga-color-surface-action);
  }
  &:checked + .marker > .checked {
    display: block;
  }
  &:indeterminate + .marker > .indeterminate {
    display: block;
  }

  &:disabled {
    cursor: not-allowed;

    + .marker {
      border-color: var(--ga-color-border-disabled);
      background-color: var(--ga-color-surface-disabled);
      color: var(--ga-color-icon-on-disabled);
    }

    ~ .label {
      color: var(--ga-color-text-disabled);
    }
  }
}

.error {
  > .marker {
    border-color: var(--ga-color-border-error);
    background-color: var(--ga-color-surface-error);
    color: var(--ga-color-icon-error);
  }

  > .native {
    &:checked + .marker,
    &:indeterminate + .marker {
      background-color: var(--ga-color-surface-error);
    }
    &:hover:enabled + .marker {
      border-color: var(--ga-color-border-error-hover);
      background-color: var(--ga-color-surface-action-hover-2);
    }
  }
}

.label {
  font-weight: var(--ga-font-weight-normal);
  font-size: var(--ga-text-md-font-size);
  line-height: var(--ga-text-md-line-height);
  font-family: var(--ga-font-family-primary);
  letter-spacing: var(--ga-text-md-tracking);

  &:empty {
    display: none;
  }
}
</style>
