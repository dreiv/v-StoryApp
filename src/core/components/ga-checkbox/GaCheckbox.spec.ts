import { shallowMount, type VueWrapper } from '@vue/test-utils'
import GaCheckbox from './GaCheckbox.vue'

vi.mock('@/core/composables/useFormInput')

describe('Checkbox', () => {
  let wrapper: VueWrapper

  beforeEach(() => {
    wrapper = shallowMount(GaCheckbox)
  })

  it('should render correctly with default props', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render correctly with label prop', () => {
    wrapper = shallowMount(GaCheckbox, { props: { label: 'Accept terms' } })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.text()).toContain('Accept terms')
  })

  it('should render correctly with default slot', () => {
    wrapper = shallowMount(GaCheckbox, { slots: { default: 'Mock Content' } })
    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.text()).toContain('Mock Content')
  })

  it('should render with state prop when provided', async () => {
    await wrapper.setProps({ label: 'Required Label', state: 'Test State' })
    expect(wrapper.text()).toContain('Test State')
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render with state slot when provided', () => {
    wrapper = shallowMount(GaCheckbox, {
      props: { label: 'Required Label' },
      slots: { state: '<div class="custom-state">Custom State</div>' },
    })
    expect(wrapper.find('.custom-state').exists()).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })

  it.each([
    { state: 'checked', props: { modelValue: true }, className: '.checked' },
    { state: 'indeterminate', props: { indeterminate: true }, className: '.indeterminate' },
  ])(
    'should render in $state state when appropriate props are set',
    async ({ props, className }) => {
      await wrapper.setProps(props)
      expect(wrapper.find(className).exists()).toBe(true)
    },
  )

  it('should emit update:modelValue event when checkbox is clicked', async () => {
    const input = wrapper.find('input')
    await input.setValue(true)

    expect(wrapper.emitted('update:modelValue')).toBeTruthy()
    expect(wrapper.emitted('update:modelValue')?.[0]).toEqual([true])
  })

  it('should toggle the checkbox state when clicked multiple times', async () => {
    const input = wrapper.find('input')

    await input.setValue(true)
    expect(wrapper.emitted('update:modelValue')?.[0]).toEqual([true])

    await input.setValue(false)
    expect(wrapper.emitted('update:modelValue')?.[1]).toEqual([false])
  })

  it('should render with error styling when error prop is true', async () => {
    await wrapper.setProps({ error: true, errorMessage: 'Test Error' })
    const label = wrapper.find('label')

    expect(wrapper.html()).toMatchSnapshot()
    expect(label.classes('mock-error')).toBe(true)
    expect(label.attributes('aria-invalid')).toBe('true')
    expect(label.attributes('aria-errormessage')).toBe('Test Error')
  })

  it('should be disabled when disabled prop is true', async () => {
    await wrapper.setProps({ disabled: true })
    const input = wrapper.find('input')

    expect(input.element.disabled).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
