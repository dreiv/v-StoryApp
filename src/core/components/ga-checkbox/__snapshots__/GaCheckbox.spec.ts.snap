// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Checkbox > should be disabled when disabled prop is true 1`] = `
"<label class="mock-input"><input type="checkbox" class="native" disabled="">
  <div class="marker">
    <anonymous-stub class="checked" size="12" stroke-width="4"></anonymous-stub>
    <anonymous-stub class="indeterminate" size="12" stroke-width="4"></anonymous-stub>
  </div>
  <!--v-if-->
</label>"
`;

exports[`Checkbox > should render correctly with default props 1`] = `
"<label class="mock-input"><input type="checkbox" class="native">
  <div class="marker">
    <anonymous-stub class="checked" size="12" stroke-width="4"></anonymous-stub>
    <anonymous-stub class="indeterminate" size="12" stroke-width="4"></anonymous-stub>
  </div>
  <!--v-if-->
</label>"
`;

exports[`Checkbox > should render correctly with default slot 1`] = `
"<label class="mock-input"><input type="checkbox" class="native">
  <div class="marker">
    <anonymous-stub class="checked" size="12" stroke-width="4"></anonymous-stub>
    <anonymous-stub class="indeterminate" size="12" stroke-width="4"></anonymous-stub>
  </div>
  <div class="label"><span class="text">Mock Content</span>
    <!--v-if-->
  </div>
</label>"
`;

exports[`Checkbox > should render correctly with label prop 1`] = `
"<label class="mock-input"><input type="checkbox" class="native">
  <div class="marker">
    <anonymous-stub class="checked" size="12" stroke-width="4"></anonymous-stub>
    <anonymous-stub class="indeterminate" size="12" stroke-width="4"></anonymous-stub>
  </div>
  <div class="label"><span class="text">Accept terms</span>
    <!--v-if-->
  </div>
</label>"
`;

exports[`Checkbox > should render with error styling when error prop is true 1`] = `
"<label class="mock-input mock-error" aria-invalid="true" aria-errormessage="Test Error"><input type="checkbox" class="native">
  <div class="marker">
    <anonymous-stub class="checked" size="12" stroke-width="4"></anonymous-stub>
    <anonymous-stub class="indeterminate" size="12" stroke-width="4"></anonymous-stub>
  </div>
  <!--v-if-->
</label>"
`;
